sigmastate/src/main/scala/sigmastate/PrettyPrintErgoTree.scala

```diff
@@ -0,0 +1,318 @@
+    createDoc(input) + Doc.text(s".$name") + argsDoc
+  }
+  
+  private def nodeWithPriorityParens(v: SValue)(implicit indent: Int): Doc = v match {
+    case _:BinAnd | _:BinOr | _:BinXor | _:SimpleRelation[_] | _:LogicalNot => wrapWithParens(createDoc(v))
+    case _ => createDoc(v)
+  }
+
+  private def binOpWithPriorityParens(l: SValue, r: SValue, sep: Doc)(implicit indent: Int): Doc =
+    nodeWithPriorityParens(l) + sep + nodeWithPriorityParens(r)
+
+  private def wrapWithParens(d: Doc)(implicit indent: Int): Doc = d.tightBracketBy(Doc.char('('), Doc.char(')'), indent)
+  private def wrapWithBrackets(d: Doc)(implicit indent: Int): Doc = d.tightBracketBy(Doc.char('['), Doc.char(']'), indent)
+
+  /* Create argument representation enclosed in brackets with types, e.g. `(str5: String, i1: Int)` */
+  private def argsWithTypesDoc(args: Seq[(Int, SType)])(implicit indent: Int): Doc = {
+    val argsWithTypes = args.map { 
+      case (i, tpe) => 
+        val argName = inferNameFromType(tpe)
+        Doc.text(s"$argName$i: ") + STypeDoc(tpe)
+    }
+    nTupleDoc(argsWithTypes)
+  }
+
+  private def STypeDoc(tpe: SType)(implicit indent: Int): Doc = tpe match {
+    case SBoolean => Doc.text("Boolean")
+    case SByte => Doc.text("Byte")
+    case SShort => Doc.text("Short")
+    case SInt => Doc.text("Int")
+    case SLong => Doc.text("Long")
+    case SBigInt => Doc.text("BigInt")
+    case SBox => Doc.text("Box")
+    case SSigmaProp => Doc.text("SigmaProp")
+    case SCollectionType(elemType) => Doc.text("Coll") + wrapWithBrackets(STypeDoc(elemType))
+    case STuple(items) => nTupleDoc(items.map(STypeDoc))
+    case SContext => Doc.text("Context")
+    case SOption(elemType) => Doc.text("Option") + wrapWithBrackets(STypeDoc(elemType))
+    case SGroupElement => Doc.text("GroupElement")
+    case SPreHeader => Doc.text("PreHeader")
+    case SString => Doc.text("String")
+    case SAny => Doc.text("Any")
+    case SAvlTree => Doc.text("AvlTree")
+    case SUnit => Doc.text("Unit")
+    case SHeader => Doc.text("Header")
+    // TODO: Not tested
+    case SGlobal => Doc.text("Global")
+    case SFunc(tDom, tRange, tpeParams) => nTupleDoc(tDom.map(STypeDoc)) + Doc.text(" => ") + STypeDoc(tRange)
+    // TODO: Are all nodes replaced after bind/typing phase? So it cannot be part of final tree?
+    case NoType => Doc.empty
+    // Not used in final ergo tree
+    case STypeApply(name, args) => ???
+    case STypeVar(name) => ???
+  }
+
+  // Create `(item1, item2, ...)`
+  private def nTupleDoc(items: Seq[Doc])(implicit indent: Int): Doc =
+    wrapWithParens(
+      Doc.intercalate(Doc.text(", "), items))
+
+  private def inferNameFromType(tpe: SType): String = tpe match {
+    case SBoolean => "bool"
+    case SByte => "b"
+    case SShort => "s"
+    case SInt => "i"
+    case SLong => "l"
+    case SBigInt => "bi"
+    case SBox => "box"
+    case SSigmaProp => "prop"
+    case SCollectionType(elemType) => "coll"
```

Comment:
elemType is not used, please use _ for unused positions here and in other patterns